README.md
# Percona MongoDB Exporter

[![Release](https://img.shields.io/github/release/percona/mongodb_exporter.svg?style=flat)](https://github.com/jpgpereira/mongodb_exporter/releases/latest)
[![Build Status](https://travis-ci.com/percona/mongodb_exporter.svg?branch=master)](https://travis-ci.com/percona/mongodb_exporter)
[![codecov.io Code Coverage](https://img.shields.io/codecov/c/github/percona/mongodb_exporter.svg?maxAge=2592000)](https://codecov.io/github/percona/mongodb_exporter?branch=master)
[![Go Report Card](https://goreportcard.com/badge/github.com/jpgpereira/mongodb_exporter)](https://goreportcard.com/report/github.com/jpgpereira/mongodb_exporter)
[![CLA assistant](https://cla-assistant.percona.com/readme/badge/percona/mongodb_exporter)](https://cla-assistant.percona.com/percona/mongodb_exporter)

Based on [MongoDB exporter](https://github.com/dcu/mongodb_exporter) by David Cuadrado ([@dcu](https://github.com/dcu)), but forked for full sharded support and structure changes.

## Features

- MongoDB Server Status metrics (*cursors, operations, indexes, storage, etc*)
- MongoDB Replica Set metrics (*members, ping, replication lag, etc*)
- MongoDB Replication Oplog metrics (*size, length in time, etc*)
- MongoDB Sharding metrics (*shards, chunks, db/collections, balancer operations*)
- MongoDB RocksDB storage-engine metrics (*levels, compactions, cache usage, i/o rates, etc*)
- MongoDB WiredTiger storage-engine metrics (*cache, blockmanger, tickets, etc*)
- MongoDB Top Metrics per collection (writeLock, readLock, query, etc*)

### Important Note

Metrics `mongodb_mongod_replset_oplog_*` doesn't work in [Master/Slave](https://docs.mongodb.com/v3.4/core/master-slave/) replication mode, because it was *DEPRECATED* in MongoDB `3.2` and removed in `4.0`.

## Building and running

### Prerequisites:

* [Go compiler](https://golang.org/dl/)
* Docker and [Docker Compose](https://docs.docker.com/compose/)

### Building

1. Get the code from the Percona repository:
 
    ```bash
    go get -u github.com/jpgpereira/mongodb_exporter
    ```
 2. Switch to the buld directory and just run ``make`` to install all needed tools, format code with `go fmt`, build a binary for your OS and run tests.:
 
    ```bash
    cd ${GOPATH-$HOME/go}/src/github.com/jpgpereira/mongodb_exporter
    make
    ```
    *Note: Running tests requires ``docker`` (as it uses MongoDB) and ``docker-compose``, and you will also need free ``27017`` port, as ``docker-compose`` maps this port into your host OS while testing.*

    1. If you want just build a binary for your OS without codestyle checks and tests you can run command below:

       ```bash
          make build
       ```

    2. If you don't have or don't want to install the whole GO stuff, use this docker build that creates a container with a freshly built `mongodb_exporter` binary:

       ```bash
          make docker
       ```

### Running

To define your own MongoDB URL, use environment variable `MONGODB_URI`. If set this variable takes precedence over `--mongodb.uri` flag.

To enable HTTP basic authentication, set environment variable `HTTP_AUTH` to user:password pair. Alternatively, you can
use YAML file with `server_user` and `server_password` fields.

```bash
export MONGODB_URI='mongodb://localhost:27017'
export HTTP_AUTH='user:password'
./bin/mongodb_exporter [<flags>]
```

#### Kubernetes

You can use the chart [prometheus-mongodb-exporter](https://github.com/helm/charts/tree/master/stable/prometheus-mongodb-exporter) from helm stable repository.

### Flags

See the help page with `-h`.

If you use [MongoDB Authorization](https://docs.mongodb.org/manual/core/authorization/), you must:

1. Create a user with '*clusterMonitor*' role and '*read*' on the '*local*' database, like the following (*replace username/password!*):

```js
db.getSiblingDB("admin").createUser({
    user: "mongodb_exporter",
    pwd: "s3cr3tpassw0rd",
    roles: [
        { role: "clusterMonitor", db: "admin" },
        { role: "read", db: "local" }
    ]
})
```

2. Set environment variable `MONGODB_URI` before starting the exporter:

```bash
export MONGODB_URI=mongodb://mongodb_exporter:s3cr3tpassw0rd@localhost:27017
```

If you use [x.509 Certificates to Authenticate Clients](https://docs.mongodb.com/manual/tutorial/configure-x509-client-authentication/), pass in username and `authMechanism` via [connection options](https://docs.mongodb.com/manual/reference/connection-string/#connections-connection-options) to the MongoDB uri. Eg:

```
mongodb://CN=myName,OU=myOrgUnit,O=myOrg,L=myLocality,ST=myState,C=myCountry@localhost:27017/?authMechanism=MONGODB-X509
```

## Note about how this works

Point the process to any mongo port and it will detect if it is a mongos, replicaset member, or stand alone mongod and return the appropriate metrics for that type of node. This was done to prevent the need to an exporter per type of process.

## Roadmap

- Document more configurations options here
- Stabilize RocksDB and WiredTiger support
- Move MongoDB user/password/authdb to a file (for security)
- Write more go tests


## Submitting Bug Reports

If you find a bug in Percona MongoDB Exporter or one of the related projects, you should submit a report to that project's [JIRA](https://jira.percona.com) issue tracker.

Your first step should be [to search](https://jira.percona.com/issues/?jql=project=PMM%20AND%20component=MongoDB_Exporter) the existing set of open tickets for a similar report. If you find that someone else has already reported your problem, then you can upvote that report to increase its visibility.

If there is no existing report, submit a report following these steps:

1. [Sign in to Percona JIRA.](https://jira.percona.com/login.jsp) You will need to create an account if you do not have one.
2. [Go to the Create Issue screen and select the relevant project.](https://jira.percona.com/secure/CreateIssueDetails!init.jspa?pid=11600&issuetype=1&priority=3&components=11603)
3. Fill in the fields of Summary, Description, Steps To Reproduce, and Affects Version to the best you can. If the bug corresponds to a crash, attach the stack trace from the logs.

An excellent resource is [Elika Etemad's article on filing good bug reports.](http://fantasai.inkedblade.net/style/talks/filing-good-bugs/).

As a general rule of thumb, please try to create bug reports that are:

- *Reproducible.* Include steps to reproduce the problem.
- *Specific.* Include as much detail as possible: which version, what environment, etc.
- *Unique.* Do not duplicate existing tickets.
- *Scoped to a Single Bug.* One bug per report.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Changed
- [PMM-4719](https://jira.percona.com/browse/PMM-4719): Remove redundant flags from "mongodb_exporter" if possible. 
Those flags have been removed: `--mongodb.authentification-database, --mongodb.max-connections, --mongodb.socket-timeout, --mongodb.sync-timeout`. You can use [connection-string-options](https://docs.mongodb.com/manual/reference/connection-string/#connection-string-options) instead.

### Added

### Fixed
- [PMM-2717](https://jira.percona.com/browse/PMM-2717): Failed to execute find query on 'config.locks': not found. source="sharding_status.go:106". 
All `mongodb_mongos_sharding_balancer_lock_*` metrics won't be exposed for `MongoDB 3.6+`. See: https://docs.mongodb.com/v3.6/reference/config-database/#config.locks.

## [0.10.0]
### Changed
- `go.mongodb.org/mongo-driver` was updated to `v1.1.1`.
- All `--mongodb.tls*` flags were removed. Use [tls-options](https://docs.mongodb.com/manual/reference/connection-string/#tls-options) instead.

### Added

### Fixed

## [0.9.0]
### Changed

### Added
- [PMM-4131](https://jira.percona.com/browse/PMM-4131): Added missing features from [dcu/mongodb_exporter](https://github.com/dcu/mongodb_exporter). See list below.
- New metrics:
  - `mongodb_mongod_replset_member_*`
  - `mongodb_connpoolstats_*`
  - `mongodb_tcmalloc_*`
- Added application name "mongodb_exporter" to mongo logs, [@nikita-b](https://github.com/nikita-b)

### Fixed
- [PMM-4427](https://jira.percona.com/browse/PMM-4427): Panic when read rocksdb status, txh [@lijinglin2019](https://github.com/lijinglin2019).
- [PMM-4583](https://jira.percona.com/browse/PMM-4583): Fix panic when GetTotalChunksByShard (#158), txh [@lijinglin2019](https://github.com/lijinglin2019).

## [0.8.0]
### Changed
- [PMM-3511](https://jira.percona.com/browse/PMM-3511): Switched to [mongo-go-driver](https://github.com/mongodb/mongo-go-driver).
- [PMM-4168](https://jira.percona.com/browse/PMM-4168): Updated all dependencies including `github.com/prometheus/golang_client`.
- Moved `opcounters`, `opcountersRepl` to `common_collector` (#146), thx [@nikita-b](https://github.com/nikita-b).

### Added

### Fixed

## [0.7.1]
### Added
- Added Authentification Database option when connect to mongo #139, thx [@etiennecoutaud](https://github.com/etiennecoutaud).
- Added helm chart to readme #140, thx [@pgdagenais](https://github.com/pgdagenais).
- [PMM-4154](https://jira.percona.com/browse/PMM-4154): Added standard logging flags.

### Fixed
- Fixed some function comments based on best practices from Effective Go #137, thx [@CodeLingoBot](https://github.com/CodeLingoBot).
- [PMM-3473](https://jira.percona.com/browse/PMM-3473): Fixed panic and runtime error.

## [0.7.0]
### Changed
- [PMM-3512](https://jira.percona.com/browse/PMM-3512): Switched to [kingpin](https://github.com/alecthomas/kingpin) library.
This is a **BREAKING CHANGE** because kingpin uses `--` instead of `-` for long flags, so be careful when updating.
- [PMM-2261](https://jira.percona.com/browse/PMM-2261) Unify common mongod and mongos server metrics, thx [@bz2](https://github.com/bz2)
This is a **BREAKING CHANGE**. The labels of these metrics are now prefixed with just `mongodb_` rather than `mongodb_mongo[ds]_`.

### Added
- Fine grained error handling for index usage and collection stats (#128), thx [@akira-kurogane](https://github.com/akira-kurogane)
- Introduce a docker go build that creates a mongodb_exporter binary within a container (#112), thx [@mminks](https://github.com/mminks)
- Ability to make releases and snapshots with [GoReleaser](https://goreleaser.com/)

## [0.6.3] - 2019-02-13
### Added
- PMM-3401: Added collection of TTL metrics #127, thx [@fastest963](https://github.com/fastest963)
- Added some new metrics:
  - `member_replication_lag`
  - `member_operational_lag`
  - `op_latencies_latency_total`
  - `op_latencies_ops_total`
  - `op_latencies_histogram`

### Fixed
- Fix for broken labels on `index_usage` metrics
- Fix SIGSEGV when running connected to a Primary, thx [@anas-aso](https://github.com/anas-aso)

### Changed
- Move CHANGELOG.md to "Keep a Changelog" format.

## [0.6.2] - 2018-09-11
### Added
- Build binaries #110
- Test `--help` flag for diff of options between releases #111

## [0.6.1] - 2018-06-15
### Fixed
- `--version` now properly reports `0.6.1`

## [0.6.0] - 2018-06-13
### Added
- Add required testify dependency #107, thx [@RubenHoms](https://github.com/RubenHoms)
- Add timeout flags #100, thx [@unguiculus](https://github.com/unguiculus)
- Enable collection of table top metrics #94, thx [@bobera](https://github.com/bobera)
- Individual index usage stats and index sizes added #97, thx [@martinhoefling](https://github.com/martinhoefling)
- Support `&ssl=true` #105 #90, thx [@dbmurphy](https://github.com/dbmurphy)

### Fixed
- Fix `balancerIsEnabled` & `balancerChunksBalanced` values #106, thx [@jmsantorum](https://github.com/jmsantorum)

## [0.5.0] - 2018-05-24
### Added
- Check connection with exporter #92. Adds `--test` flag to verify connection with MongoDB and quits.

### Removed
- Removed tests for EOL'ed MongoDB 3.0 and Percona Server for MongoDB 3.0.

### Fixed
- Redact Mongo URI #101. Fixes URI logging in plain text including credentials when no session can be created.
- ARM64-specific fixes #102. Fixed two portability issues in Makefile.

## [0.4.0] - 2018-01-17
### Added
- New flags `-collect.database` and `-collect.collection` can be used to enable collection of database and collection
  metrics. They are disabled by default.
- MongoDB connections are now kept between the scrapes. New flag `-mongodb.max-connections` (with the default value `1`)
controls the maximum number of established connections.
- Add standard metrics:

mongodb_exporter.go
// Copyright 2017 Percona LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
	"fmt"
	"os"
	"strconv"
	"time"

	"github.com/percona/exporter_shared"
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
	"github.com/prometheus/common/log"
	"github.com/prometheus/common/version"
	"gopkg.in/alecthomas/kingpin.v2"

	pmmVersion "github.com/percona/pmm/version"

	"github.com/jpgpereira/mongodb_exporter/collector"
	"github.com/jpgpereira/mongodb_exporter/shared"
)

const (
	program           = "mongodb_exporter"
	versionDataFormat = "20060102-15:04:05"
)

var (
	listenAddressF = kingpin.Flag("web.listen-address", "Address to listen on for web interface and telemetry.").Default(":9216").String()
	metricsPathF   = kingpin.Flag("web.telemetry-path", "Path under which to expose metrics.").Default("/metrics").String()

	collectDatabaseF             = kingpin.Flag("collect.database", "Enable collection of Database metrics").Bool()
	collectCollectionF           = kingpin.Flag("collect.collection", "Enable collection of Collection metrics").Bool()
	collectTopF                  = kingpin.Flag("collect.topmetrics", "Enable collection of table top metrics").Bool()
	collectIndexUsageF           = kingpin.Flag("collect.indexusage", "Enable collection of per index usage stats").Bool()
	mongodbCollectConnPoolStatsF = kingpin.Flag("collect.connpoolstats", "Collect MongoDB connpoolstats").Bool()
	collectDatabaseProfilerF     = kingpin.Flag("collect.databaseprofiler", "Enable collection of database system.profiler metrics").Bool()
	collectDatabaseCurrentOpsF   = kingpin.Flag("collect.currentop", "Enable collection of $currentOp metrics").Bool()
	databaseProfilerLookbackF    = kingpin.Flag("databaseprofiler.lookback", "Size of the system.profile scan window, in seconds").Default("60").Int64()
	databaseProfilerThresholdF   = kingpin.Flag("databaseprofiler.threshold", "Min query duration, in ms, for slow queries to count").Default("1000").Int64()

	uriF = kingpin.Flag("mongodb.uri", "MongoDB URI, format").
		PlaceHolder("[mongodb://][user:pass@]host1[:port1][,host2[:port2],...][/database][?options]").
		Default("mongodb://localhost:27017").
		Envar("MONGODB_URI").
		String()
	testF = kingpin.Flag("test", "Check MongoDB connection, print buildInfo() information and exit.").Bool()
)

func main() {
	initVersionInfo()
	log.AddFlags(kingpin.CommandLine)
	kingpin.Parse()

	if *testF {
		buildInfo, err := shared.TestConnection(
			shared.MongoSessionOpts{
				URI: *uriF,
			},
		)
		if err != nil {
			log.Errorf("Can't connect to MongoDB: %s", err)
			os.Exit(1)
		}
		fmt.Println(string(buildInfo))
		os.Exit(0)
	}

	// TODO: Maybe we should move version.Info() and version.BuildContext() to https://github.com/percona/exporter_shared
	// See: https://jira.percona.com/browse/PMM-3250 and https://github.com/jpgpereira/mongodb_exporter/pull/132#discussion_r262227248
	log.Infoln("Starting", program, version.Info())
	log.Infoln("Build context", version.BuildContext())

	programCollector := version.NewCollector(program)
	mongodbCollector := collector.NewMongodbCollector(&collector.MongodbCollectorOpts{
		URI:                       *uriF,
		CollectDatabaseMetrics:    *collectDatabaseF,
		CollectCollectionMetrics:  *collectCollectionF,
		CollectTopMetrics:         *collectTopF,
		CollectIndexUsageStats:    *collectIndexUsageF,
		CollectConnPoolStats:      *mongodbCollectConnPoolStatsF,
		CollectDatabaseCurrentOps: *collectDatabaseCurrentOpsF,
		CollectDatabaseProfiler:   *collectDatabaseProfilerF,
		DatabaseProfilerLookback:  *databaseProfilerLookbackF,
		DatabaseProfilerThreshold: *databaseProfilerThresholdF,
	})
	prometheus.MustRegister(programCollector, mongodbCollector)

	promHandler := promhttp.InstrumentMetricHandler(prometheus.DefaultRegisterer, promhttp.HandlerFor(prometheus.DefaultGatherer, promhttp.HandlerOpts{ErrorHandling: promhttp.ContinueOnError}))
	exporter_shared.RunServer("MongoDB", *listenAddressF, *metricsPathF, promHandler)
}

// initVersionInfo sets version info
// If binary was build for PMM with environment variable PMM_RELEASE_VERSION
// `--version` will be displayed in PMM format. Also `PMM Version` will be connected
// to application version and will be printed in all logs.
// TODO: Refactor after moving version.Info() and version.BuildContext() to https://github.com/percona/exporter_shared
// See: https://jira.percona.com/browse/PMM-3250 and https://github.com/jpgpereira/mongodb_exporter/pull/132#discussion_r262227248
func initVersionInfo() {
	version.Version = pmmVersion.Version
	version.Revision = pmmVersion.FullCommit
	version.Branch = pmmVersion.Branch

	if buildDate, err := strconv.ParseInt(pmmVersion.Timestamp, 10, 64); err != nil {
		version.BuildDate = time.Unix(0, 0).Format(versionDataFormat)
	} else {
		version.BuildDate = time.Unix(buildDate, 0).Format(versionDataFormat)
	}

	if pmmVersion.PMMVersion != "" {
		version.Version += "-pmm-" + pmmVersion.PMMVersion
		kingpin.Version(pmmVersion.FullInfo())
	} else {
		kingpin.Version(version.Print(program))
	}

	kingpin.HelpFlag.Short('h')
	kingpin.CommandLine.Help = fmt.Sprintf("%s exports various MongoDB metrics in Prometheus format.\n", pmmVersion.ShortInfo())
}
